### PR TITLE
feat: refactor canary lab to use NGINX as an ingress

### PR DESCRIPTION
**Pull request by 4141done**
### Proposed changes

Basic example of a dockerized deploy of nginx with:
* Health checks at `/health`
* Fake upstreams with various failure modes

#### Quick NGINX reference for local testing
from `./load_balancer`

####### Starting the server with the config:
`nginx -c $(pwd)/nginx.conf`

####### Testing configuration to see if it will run:
`nginx -c $(pwd)/nginx.conf -t`

####### Reloading configuration when server is running:
`nginx -c $(pwd)/nginx.conf -s reload`

####### Stop NGINX:
`nginx -s quit`

Ref: https://docs.nginx.com/nginx/admin-guide/basic-functionality/runtime-control/

Best doc on nginx config: https://www.digitalocean.com/community/tutorials/understanding-the-nginx-configuration-file-structure-and-configuration-contexts

#### Deploy
When paired with a simple `Dockerfile` like this:
```
FROM nginx:1.23
COPY nginx.conf /etc/nginx/nginx.conf
```

It can be deployed on its own

### CODE REVIEW

File: load_balancer/nginx.conf [hunk 1]
```diff
@@ -10,15 +10,32 @@ http {
...
+        }
+        
+        location /health {
+          access_log off;
+          add_header 'Content-Type' 'text/plain';
+          return 200 "OK";
         }
     }
+    
+    server {
```
Comment (4141done) [inline load_balancer/nginx.conf]: **Annotation**:
This is a fake "upstream" server that just returns text from the root and provides a health check endpoint.

File: load_balancer/nginx.conf [hunk 2]
```diff
@@ -10,15 +10,32 @@ http {
...
-      server messenger:4000;
-    }
+
     server {
         listen       80;
-        server_name  localhost;
+
         location / {
-          proxy_pass http://messenger_upstream;
+          proxy_pass http://localhost:8080;
```
Comment (4141done) [inline load_balancer/nginx.conf]: **Annotation**:
This is the directive that forwards a request to `*:80` to the handler in the other server.

File: load_balancer/nginx.conf [hunk 3]
```diff
@@ -10,15 +10,32 @@ http {
...
+        }
+        
+        location /health {
+          access_log off;
+          add_header 'Content-Type' 'text/plain';
+          return 200 "OK";
```
Comment (4141done) [inline load_balancer/nginx.conf]: **Annotation**:
Pretty basic health check endpoint.  We make sure these don't pollute the access log (which is a list of incoming requests).

**Review by alessfg — APPROVED** [review]

**Comment by alessfg** [comment]
The original PR has drastically changed form across multiple PRs, but in the end, we got where we wanted :)

**Pull request merged**